### DISCUSSION TITLE
No Internet in Diagnostics and Outgoing connections fail

### DISCUSSION THREAD

**Discussion by knightian** [dani-garcia/vaultwarden#3941] [Q&A]
![image](https://github.com/dani-garcia/vaultwarden/assets/72685601/3341bc36-f7cb-457f-8d35-a02a343c6bfc)

Websockets are working fine, everything seems to work except for anything that vaultwarden needs to reach out to like version checks, NTP and push notifications.

In the log there is error setting up push:

```
[2023-10-02 00:18:49.867][vaultwarden::api::push][ERROR] Error getting push token from bitwarden server: error sending request for url (https://identity.bitwarden.com/connect/token): error trying to connect: dns error: No connections available
[2023-10-02 00:18:49.868][vaultwarden::api::core::accounts][ERROR] An error occured while proceeding registration of a device: {"ErrorModel":{"Message":"Error getting push token from bitwarden server: error sending request for url (https://identity.bitwarden.com/connect/token): error trying to connect: dns error: No connections available","Object":"error"},"ExceptionMessage":null,"ExceptionStackTrace":null,"InnerExceptionMessage":null,"Message":"Error getting push token from bitwarden server: error sending request for url (https://identity.bitwarden.com/connect/token): error trying to connect: dns error: No connections available","Object":"error","ValidationErrors":{"":["Error getting push token from bitwarden server: error sending request for url (https://identity.bitwarden.com/connect/token): error trying to connect: dns error: No connections available"]},"error":"","error_description":""}
[2023-10-02 00:18:49.868][response][INFO] (put_device_token) PUT /api/devices/identifier/<uuid>/token => 400 Bad Request
```

I can curl https://identity.bitwarden.com/connect/token from the server and it reaches it without issue. This all used to work fine it is just the last few Bitwarden versions this seems to happen, never had any issues in the past.

SMTP reaches out no worries to send emails through Microsoft 365.

Using nginx as reverse proxy, was using my own config but I have switched to the nginx config provided by @BlackDex in the proxy examples.

FWIW I also see this in the vaultwarden log:
```[2023-10-02 00:59:53.441][trust_dns_resolver::system_conf::unix][WARN] no nameservers found in config```

But DNS is working fine on the server.

I suspect this is going to be a DNS issue within Vaultwarden, but why does it only pop up in current versions did something change?

Here is the debug string:

### Your environment (Generated via diagnostics page)
* Vaultwarden version: v1.29.2-bc26bfa5
* Web-vault version: v2023.8.2
* OS/Arch: linux/x86_64
* Running within Docker: false (Base: Not applicable)
* Environment settings overridden: false
* Uses a reverse proxy: true
* IP Header check: true (X-Real-IP)
* Internet access: false
* Internet access via a proxy: false
* DNS Check: true
* Browser/Server Time Check: true
* Server/NTP Time Check: n/a
* Domain Configuration Check: true
* HTTPS Check: true
* Database type: SQLite
* Database version: 3.41.2
* Clients used: 
* Reverse proxy and version: 
* Other relevant information: 

### Config (Generated via diagnostics page)
<details><summary>Show Running Config</summary>

**Environment settings which are overridden:** 

```json
{
  "_duo_akey": null,
  "_enable_duo": false,
  "_enable_email_2fa": true,
  "_enable_smtp": true,
  "_enable_yubico": true,
  "_icon_service_csp": "",
  "_icon_service_url": "",
  "_ip_header_enabled": true,
  "_smtp_img_src": "cid:",
  "admin_ratelimit_max_burst": 2,
  "admin_ratelimit_seconds": 6,
  "admin_session_lifetime": 20,
  "admin_token": "***",
  "allowed_iframe_ancestors": "",
  "attachments_folder": "data/attachments",
  "auth_request_purge_schedule": "30 * * * * *",
  "authenticator_disable_time_drift": false,
  "data_folder": "data",
  "database_conn_init": "",
  "database_max_conns": 256,
  "database_timeout": 30,
  "database_url": "***************",
  "db_connection_retries": 20,
  "disable_2fa_remember": false,
  "disable_admin_token": false,
  "disable_icon_download": false,
  "domain": "*****://********************",
  "domain_origin": "*****://********************",
  "domain_path": "",
  "domain_set": true,
  "duo_host": null,
  "duo_ikey": null,
  "duo_skey": null,
  "email_attempts_limit": 5,
  "email_expiration_time": 1800,
  "email_token_size": 6,
  "emergency_access_allowed": false,
  "emergency_notification_reminder_schedule": "0 3 * * * *",
  "emergency_request_timeout_schedule": "0 7 * * * *",
  "enable_db_wal": true,
  "event_cleanup_schedule": "0 10 0 * * *",
  "events_days_retain": null,
  "extended_logging": true,
  "helo_name": null,
  "hibp_api_key": null,
  "icon_blacklist_non_global_ips": true,
  "icon_blacklist_regex": null,
  "icon_cache_folder": "data/icon_cache",
  "icon_cache_negttl": 259200,
  "icon_cache_ttl": 2592000,
  "icon_download_timeout": 10,
  "icon_redirect_code": 302,
  "icon_service": "internal",
  "incomplete_2fa_schedule": "30 * * * * *",
  "incomplete_2fa_time_limit": 3,
  "invitation_expiration_hours": 120,
  "invitation_org_name": "Mine",
  "invitations_allowed": true,
  "ip_header": "X-Real-IP",
  "job_poll_interval_ms": 30000,
  "log_file": "/var/log/vaultwarden/vaultwarden.log",
  "log_level": "info",
  "log_timestamp_format": "%Y-%m-%d %H:%M:%S.%3f",
  "login_ratelimit_max_burst": 3,
  "login_ratelimit_seconds": 4,
  "org_attachment_limit": null,
  "org_creation_users": "**********************",
  "org_events_enabled": false,
  "org_groups_enabled": false,
  "password_hints_allowed": false,
  "password_iterations": 700442,
  "push_enabled": true,
  "push_installation_id": "***",
  "push_installation_key": "***",
  "push_relay_uri": "https://push.bitwarden.com",
  "reload_templates": false,
  "require_device_email": true,
  "rsa_key_filename": "data/rsa_key",
  "send_purge_schedule": "0 5 * * * *",
  "sendmail_command": null,
  "sends_allowed": true,
  "sends_folder": "data/sends",
  "show_password_hint": false,
  "signups_allowed": false,
  "signups_domains_whitelist": "***********,********************",
  "signups_verify": true,
  "signups_verify_resend_limit": 20,
  "signups_verify_resend_time": 300,
  "smtp_accept_invalid_certs": false,
  "smtp_accept_invalid_hostnames": false,
  "smtp_auth_mechanism": "\"Login\"",
  "smtp_debug": false,
  "smtp_embed_images": true,
  "smtp_explicit_tls": null,
  "smtp_from": "*****************************",
  "smtp_from_name": "Mine",
  "smtp_host": "******************",
  "smtp_password": "***",
  "smtp_port": 587,
  "smtp_security": "starttls",
  "smtp_ssl": null,
  "smtp_timeout": 15,
  "smtp_username": "***********************************",
  "templates_folder": "data/templates",
  "tmp_folder": "data/tmp",
  "trash_auto_delete_days": 30,
  "trash_purge_schedule": "0 5 0 * * *",
  "use_sendmail": false,
  "use_syslog": false,
  "user_attachment_limit": 102400,
  "web_vault_enabled": true,
  "web_vault_folder": "web-vault/",
  "websocket_address": "0.0.0.0",
  "websocket_enabled": false,
  "websocket_port": 3012,
  "yubico_client_id": null,
  "yubico_secret_key": null,
  "yubico_server": null
}
```
</details>

**Comment by knightian** [comment]
I solved this problem.

I looked at the code and I see that vaultwarden is using a package called `trust-dns-resolver`

I found this article: https://crates.io/crates/trust-dns-resolver and I noticed that trust-dns-resolver package is parsing `/etc/resolv.conf` looking for nameserver entries. Because I am receiving my DNS via DHCP options, I didn't have any nameserver entries in this file and so it was causing Bitwarden to not be able to do any DNS lookups.

I added:

`nameserver 127.0.0.1` into /etc/resolv.conf because my server also runs a local unbound instance that can be used. Once I did this, it fixed the issues (but now push notifications with iOS are not working still).

**So long story short**

If you do not have any nameserver listed in `/etc/resolv.conf` it will break vaultwarden for some things (mostly outgoing requests)

**Comment by knightian** [comment]
I would like to NOT have to have a nameserver entry in `/etc/resolv.conf`, is it possible to force trust-dns-resolver to use a particular DNS server instead? Maybe through an ENV var in Vaultwarden?

**Comment by stefan0xC** [comment]
@knightian I don't think so. This also seems rather esoteric to me. And I'm not sure if we want to add another configuration option. But you should probably ask [in the trust-dns repository](https://github.com/bluejekyll/trust-dns) if this is possible or something they'd consider adding, if it's not.

**Comment by knightian** [comment]
> This also seems rather esoteric to me. 

I see, but if someone is using a VPS and the cloud provider is using Netplan to configure the network and DNS is coming from DHCP options and is not resident in resolv.conf then they are going to have the same bad time.

How to get around it? Or just allow people who install into VPS with these setups to fail?

**Comment by BlackDex** [comment]
Well, by default docker takes the `resolv.conf` from the host, if that isn't ok, you should use `docker run --dns=1.1.1.1` or it's equivalent for docker-compose.

**Reply by knightian** [reply]
If you are using docker, but I am not.

**Reply by BlackDex** [reply]
Well, if trust-dns doesn't work, it falls back to other ways, so not sure why it doesn't work then.
But maybe something changed on your host.

**Reply by knightian** [reply]
Nah I have 2 hosts (one is a failover) and the issue is on both, nothing has changed from me in a long time, unless it's a system update that has made changes which is possible. Also if "it falls back to other ways" includes DNS resolution by google or cloudflare or something (using port 53) this would not be allowed on my system but this has always been the case.

**Reply by BlackDex** [reply]
It uses Reqwest's default DNS handling, which is handled by the system it self, which is normally a system api call.
The strange thing though, it says DNS works, because it can resolve `github.com` as you can see in the diagnostics page, so it actually looks like DNS does seem to work in some way for Vaultwarden, but just no internet access, so in this case, no access to `github.com`.

You can see here, that if loading with trust-dns fails, we use the default:
https://github.com/dani-garcia/vaultwarden/blob/bc26bfa589c007da9b9be37e1172060f38a948b9/src/util.rs#L683-L691

**Reply by knightian** [reply]
"You can see here, that if loading with trust-dns fails, we use the default:"

Yep I saw that, and it's supposed to write an entry in the log ""Possible trust-dns error, trying with trust-dns disabled: '{e}'"" and I checked the log for these messages and they do not exist. it looks like trust-dns is trowing expection and Vaultwarden is not falling back to this anymore

If I put an entry in `/etc/resolv.conf` then all is working according to diagnostics:

![image](https://github.com/dani-garcia/vaultwarden/assets/72685601/3adca7b7-ac40-4372-9b6c-271fa963861a)

**Reply by knightian** [reply]
@BlackDex  I don't know if it matters but it seems like trust-dns treats no nameservers as a warning not an error?

trust-dns is printing `[2023-10-02 00:59:53.441][trust_dns_resolver::system_conf::unix][WARN] no nameservers found in config` but it is classing it as a warning, so it might not fit the criteria of Err to hit that fallback condition?

There are no log entries that state it is falling back as there should be if it is, so I suspect what is happening is when trust-dns is not detecting nameservers, it is not throwing error it is just doing warning only, so the fallback code is not reached.

**Reply by stefan0xC** [reply]
Just to clarify, you do have a `/etc/resolv.conf` file? Just without any entries?

**Reply by knightian** [reply]
@stefan0xC 
it had a search entry
so:

`search domain.com`

But it did not have a nameserver entry.

I have added a nameserver entry `nameserver 127.0.0.1` and it makes vaultwarden work in doing so. It seems the issue is that trust-dns regards no nameservers in resolv.conf as a warning not an error, and then vaultwarden does not fall back to other DNS alternatives in this case.

**Reply by stefan0xC** [reply]
Did you add `search domain.com` manually or was this file created automatically?

**Comment by knightian** [comment]
> Just to clarify, you do have a `/etc/resolv.conf` file? Just without any entries?

@stefan0xC yes so like in this article https://askubuntu.com/questions/1292650/resolv-conf-no-dns-servers-known

"Ubuntu 20.04 uses /etc/netplan to set DNS" and "resolve.conf is generated from netplan during boot"

My .yml file issued by the cloud provider for netplan to setup the network is not specifying nameserver, so it is not putting nameserver into resolv.conf

Instead it is letting the DNS servers come through DHCP options, so when it makes DHCP request it gets the DNS server in that response, where it is caching them I don't know, but it is not resolv.conf I know that much.

This is the .yaml fie if it is interesting:

```
# This file is generated from information provided by the datasource.  Changes
# to it will not persist across an instance reboot.  To disable cloud-init's
# network configuration capabilities, write a file
# /etc/cloud/cloud.cfg.d/99-disable-network-config.cfg with the following:
# network: {config: disabled}
network:
    ethernets:
        ens3:
            dhcp4: true
            dhcp6: true
            match:
                macaddress: xx:xx:xx:xx:xx:xx
            set-name: ens3
    version: 2

**Reply by stefan0xC** [reply]
> where it is caching them I don't know, but it is not resolv.conf I know that much.

This could be the issue then.

It probably depends what program you are using, i.e. `/etc/resolv.conf` should be a symlink to `../run/systemd/resolve/stub-resolv.conf` if you're using `systemd-resolved` (which should be default since `17.10`). Or to `/run/resolvconf/resolv.conf` if you are using `resolvconf`. Otherwise I think `NetworkManager` would maybe add the entries to `/etc/resolv.conf` directly?

**Reply by BlackDex** [reply]
I think there's is a system configuration issue which could also cause other applications to have problems with DNS. Maybe not much, but still.

Reading this seems that this happens often and it also provides solutions on how to solve this. 
https://askubuntu.com/questions/1135486/make-netplan-write-resolve-conf

If I'm correct netplan needs systemd-resolved as @stefan0xC pointed. And that file should always contain 127.0.0.53 as a nameserver.

**Reply by knightian** [reply]
@BlackDex  "I think there's is a system configuration issue which could also cause other applications to have problems with DNS. Maybe not much, but still."

There is no problem with DNS on my server. Actually I am running unbound on it, so the server itself is a DNS server, thus why 127.0.0.1 resolves DNS. systemd-resolve is not present on the server, but this may be due to me installing unbound, although I did this years ago, end of 2020.

@stefan0xC I believe the search entry was put in there by the cloud provider, it is on all 4 VPS instances and I don't recall putting it in there and I do know in the cloud providers panel I can specify the domain so I assume it's putting it in there when provisioning the image. If I did manually do it, it would have been 2 years ago or even more, and this ties into another point, I have been using Vaultwarden for years, I was using it on these same servers when it was called Bitwarden_rs, and the diagnostic page never saw these issues. Only in the last few versions this started happening, well, anytime from June this year to now it could have happened.

So we know that trust-dns checks resolv.conf for nameservers and if none are present it fails. I have fixed this as you know by putting nameserver entry to my unbound instance running on the device at 127.0.0.1, but this is NEW behaviour, this has only started some time since June this year. The DNS issue that ONLY vaultwarden has (through the use of trust-dns) has been resolved by this action, there is no DNS issues on this device, don't make the mistake of thinking that because both trust-dns and vaultwarden are failing to fallback to trying to resolve using the system DNS that there is system DNS issue, that is 100% not the case here.

If I just "nslookup domain.com" the system resolves names just fine, and therefore anything using the system to resolve names does so equally fine. Again, this is a trust-dns issue, and then an issue with Vaultwarden where it does not recover properly when trust-dns fails, it is not falling back when trust-dns fails, your code to fall back is not being reached in this instance and it is not writing the log entries so we know this is a fact.

I have fixed this issue for me, as we know, but the question is as you say @BlackDex "this happens often" and so my line of thinking is that if it happens often, do you want Vaultwarden to fail often in this way? But then again, if you are using the docker image as you also say it should not be an issue because resolv.conf will be set in the docker image and docker will just resolve from the system. So, if it is the case that the expectation is that everyone uses docker, then this may be a non-issue in that case. But, if you wanted to make sure this worked all the time without docker, then currently Vaultwarden code is buggy in that it is supposed to fallback to a different DNS when trust-dns fails, and currently it is not. We know it is not, because the log entries that are part of that code are not being written.

It's up to code maintainers to fix the code so it actually falls back to system resolution when trust-dns fails, or just say it's not an issue and ignore it, I have fixed it for me so it's up to you how you want to handle it I guess.

**Reply by BlackDex** [reply]
> I have fixed this issue for me, as we know, but the question is as you say @BlackDex "this happens often" and so my line of thinking is that if it happens often, do you want Vaultwarden to fail often in this way?

The happens often reverse to the article i posted. I didn't say it happens often for Vaultwarden. As far as i can tell you are the very first.
Still, i think the resolv.conf should either not exists, or have a nameserver located in there.

You say it's a code bug, but i'm not sure. If the resolv.conf needs a nameserver (which it does as far as i know) then it is more a system problem. But i have not looked into that, i have a lot of other stuff to check and finish first before i will have some time to check this specific item. Since you already have it fixed right now it should be ok.

**Reply by BlackDex** [reply]
Also, if other people want to dive into to this, and check how resolv.conf should be configured or not exists, then that is also fine, don't let me stop others to check this hehe.

**Reply by knightian** [reply]
@BlackDex "You say it's a code bug, but i'm not sure." The reason I say it's a code bug is because failover from trusrt-dns is supposed to occur, and it is not. If the code just ran trust-dns and did not contain any logic to failover should trust-dns fail then I could accept it to be just a system issue, but given that the code to failover is not actually failing over, this is a logic error in the code. I reckon 99% of people will be using docker and even those who don't this is agreeably a reasonably unique circumstance, but, it does beg the question of if you are putting in code to failover, you would be wanting it to failover right?

**Reply by stefan0xC** [reply]
@knightian If your VPS replaces the file with a faulty configuration it's an issue you should report to your VPS. I don't think it happens often to our users (at least not that I know of) but (indicated by [the many questions on askubuntu.com](https://askubuntu.com/search?q=resolvconf)) it seems to be an issue with Ubuntu (because apparently installing `resolvconf` or `openresolv` will conflict with the default by replacing the symlink to the stub-resolver and also ubuntu's documentation seems a bit lacking).

But I agree that the behavior of the `trust-dns` feature could be improved but after looking at it a bit I'm not sure if we **can** fix the fallback behavior because neither [`reqwest::Client`](https://docs.rs/reqwest/latest/reqwest/struct.Client.html) nor [`reqwest::ClientBuilder`](https://docs.rs/reqwest/latest/reqwest/struct.ClientBuilder.html) expose a way to test for empty nameserver configuration. This is probably an issue with either the `trust-dns-resolver` (which relies on a working `/etc/resolv.conf`) or `reqwest` (which should not accept a `TrustDnsResolver` with no configured nameservers).

**Reply by BlackDex** [reply]
Well, i know @dani-garcia was working on a bit more secure DNS resolving for the favicons, see https://github.com/dani-garcia/vaultwarden/tree/icons_dns, but i also know he doesn't have much time hehe.

That might also be an option to use for other locations, but not sure.

**Comment by knightian** [comment]
> But I agree that the behavior of the `trust-dns` feature could be improved but after looking at it a bit I'm not sure if we **can** fix the fallback behavior because neither [`reqwest::Client`](https://docs.rs/reqwest/latest/reqwest/struct.Client.html) nor [`reqwest::ClientBuilder`](https://docs.rs/reqwest/latest/reqwest/struct.ClientBuilder.html) expose a way to test for empty nameserver configuration. This is probably an issue with either the `trust-dns-resolver` (which relies on a working `/etc/resolv.conf`) or `reqwest` (which should not accept a `TrustDnsResolver` with no configured nameservers).

I can see that vaultwarden is writing the warning parsed from trust-dns to logs, can't you just catch that warning and instead of simply writing to the log, then perform the necessary failover option (resolve using system DNS), would that not be the fix? @stefan0xC

**Reply by BlackDex** [reply]
That warning is logged by the crate it self, not by Vaultwarden. It only catches the log and passes it on.
There might be a way to do some checking our self. But it is not going to be on my priority list for now to be honest.
Again, i would first like that someone who can code in Rust (it might be me, but others are fine too) to check what happens during your specific case example. And also what should be in the resolv.conf file and what should happen if there are now nameservers located there, as in my experience, a lot of stuff breaks if there arn't nameservers located there.

**Reply by stefan0xC** [reply]
@knightian I think that would be a band-aid at most. The fix was to repair the broken (or at least [not recommend](https://manpages.ubuntu.com/manpages/focal/en/man8/systemd-resolved.service.8.html#/etc/resolv.conf)) system configuration.

Since we are not the immediate consumers of the crate (which could at least check for this) and also not a lot of users should be affected I don't see a reason for us to implement a workaround in Vaultwarden. Especially one that would rely on runtime testing which seems to be the only way to tell if it does not work. 

I mean there are tons of things we could do like rewriting how DNS is resolved ourselves but in my eyes the problem is `trust-dns-resolver` not throwing an error (which might be intentional, idk) and/or `reqwest` not catching this issue (as they probably should, even if the former does not say it is an error).

**Reply by knightian** [reply]
Yes I think trust-dns not throwing an error is the key here. They have asked me to submit a PR to change that log write to an exception throw and so i will be doing that. https://github.com/bluejekyll/trust-dns/issues/2042

**Comment by knightian** [comment]
> It only catches the log and passes it on.

So if it catches the log, it can parse it yah? In my view, if trust-dns says "warning, I have no nameserers configured" then surely it is up to Vaultwarden to acknowledge and act on that warning?

Just pushing on and trying to use trust-dns to resolve things after it warns you it has no way to resolve things is faulty logic for sure.

Yes I think because most will use docker, it's not a priority. But I think it should be tidied up one day I guess.

**Reply by BlackDex** [reply]
We do not catch every log and act upon that. That is just not how it works.
I'm also not inclined into going into details on how it works.
It's as simple as, crates/libraries can send trigger a log event, and that is then handled by an other crate which then outputs that to stdout/stderr and/or a log file. There is nothing which checks to log events and then tries to act upon that.

We can act upon `Results`, which either return `Ok` or `Err`, and if we do not receive an `Err`, then we need to assume all was ok. If that was not the case, then it could be that the crate is wrong or has a bug. But again, that is something we need to investigate, if this is a server error for not having a nameserver in the resolv.conf or a application error for not treating that as a huge issue.

**Reply by knightian** [reply]
> We can act upon Results, which either return Ok or Err, and if we do not receive an Err

This seems to be the crux of it, trust-dns is logging a warning, but not returning `Err`, it sems to be treating it as `Ok` but how can it be ok when it cannot perform any name resolutions? The way around it in the short term is to parse trust-dns log at that time and catch that warning, but yes I think ideally this is a failure by trust-dns, it should return an `Err` if it cannot do the one thing it exists to do.... perform DNS name resolution....

**Reply by BlackDex** [reply]
Well, go complain at the trust-dns crate then ;), again, as i said before, it's something a Rust dev can/should check, and it's not on my prio list for now.

**Comment by unrelatedinfo** [comment]
I would just like to chime in and say I observe the same behavior.

**Comment by Ovi1kanobe** [comment]
I had the same issue and my fix was that I did not have curl installed on my machine. Once I installed it everything worked fine.
